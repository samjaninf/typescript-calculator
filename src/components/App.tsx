import { Actions } from '../redux'
import { BoundActions } from '../types'
import { State } from '../redux/reducers'
import { StyleAttribute, css } from 'glamor'
import {
  add,
  clear,
  divide,
  equals,
  getOutput,
  input,
  multiply,
  percent,
  posneg,
  subtract,
} from '../redux/modules/operations'
import { bindActionCreators } from 'redux'
import { connect } from 'preact-redux'
import { h } from 'preact'
import Buttons from './Buttons'
import Calculator from './Calculator'
import Display from './Display'
import Hero from './Hero'

type Dispatch = ( action: Actions ) => never
type AppProps = {
  actions: BoundActions,
  output: string,
}

const app: StyleAttribute = css({
  alignItems: 'center',
  display: 'flex',
  flexDirection: 'column',
})

export const App = ({ actions, output }: AppProps): JSX.Element => (
  <main className={app.toString()}>
    <Hero />
    <Calculator>
      <Display output={output} />
      <Buttons actions={actions} />
    </Calculator>
  </main>
)

const mapDispatchToProps = ( dispatch: Dispatch ) => ({
  actions: bindActionCreators(
    { add, clear, divide, equals, input, multiply, percent, posneg, subtract },
    dispatch,
  ),
})

const mapStateToProps = ( state: State ) => ({
  output: getOutput(state),
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
