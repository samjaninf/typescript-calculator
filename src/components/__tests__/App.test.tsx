import { App } from '../App'
import snapshot from './utils/snapshot'

const actions = { add: () => 1 + 1 }

describe('App', () => {
  it('renders', () => snapshot(App))
  it('passes output to <Display />', () => snapshot(App, { output: '12' }))
  it('passes actions to <Buttons />', () => snapshot(App, { actions }))
})
