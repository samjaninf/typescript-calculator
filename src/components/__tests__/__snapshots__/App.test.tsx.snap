// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App passes actions to <Buttons /> 1`] = `
"<main class=\\"css-cf8zkh\\">
  <Hero></Hero>
  <Calculator>
    <Display></Display>
    <Buttons actions=\\"[object Object]\\"></Buttons>
  </Calculator>
</main>"
`;

exports[`App passes output to <Display /> 1`] = `
"<main class=\\"css-cf8zkh\\">
  <Hero></Hero>
  <Calculator>
    <Display output=\\"12\\"></Display>
    <Buttons></Buttons>
  </Calculator>
</main>"
`;

exports[`App renders 1`] = `
"<main class=\\"css-cf8zkh\\">
  <Hero></Hero>
  <Calculator>
    <Display></Display>
    <Buttons></Buttons>
  </Calculator>
</main>"
`;
